### PR TITLE
Tensor context vars no_grad and training.

### PR DESCRIPTION
**Pull request by leopf**
Context vars for training and no_grad. Allows for:
```python
from tinygrad.helpers import Context

@Context(training=True)
def train_cifar():
    ...

@Context(no_grad=True)
def run_llama():
    ...    

# or

with Context(training=True):
    ...
    # train some model
```

Tests pass for me. I was careful, but hope I did not break any examples.

I also saw #877. I will add some tests for the context vars.

### CODE REVIEW

**Comment by leopf** [comment]
Added some tests for training and no_grad.

Testing for training behavior in dropout and batchnorm2d. Also testing if no_grad actually results in there being no grad.

**Comment by tinyb0t** [comment]
Changes made in `tinygrad/`:
```
------------------------------------------------------------
files                             insertions       deletions
------------------------------------------------------------
tinygrad/helpers.py                        8               7
tinygrad/nn/__init__.py                    2               2
tinygrad/tensor.py                         6               5
------------------------------------------------------------
total                                     16              14
------------------------------------------------------------
lines added in the tinygrad folder: 2
```

**Comment by geohot** [comment]
Absolutely massive change I can't read.

The functionality should be introduced in a separate PR from the one that uses it.
